2_download/src/fetch_wqp_data.R

```diff
@@ -83,23 +83,49 @@ add_download_groups <- function(sitecounts_df, max_sites = 500, max_results = 25
   }
   
   # Subset 'good' sites with identifiers that can be parsed by WQP
-  sitecounts_good_ids <- sitecounts_df %>%
+  sitecounts_good_ids <- site_counts %>%
     filter(!MonitoringLocationIdentifier %in% sitecounts_bad_ids$site_id)
   
-  # Within each unique grid_id, use the cumsumbinning function from the MESS
-  # package to group sites based on the cumulative sum of results_count 
-  # across sites within the grid, resetting the download group/task number 
-  # if the number of records exceeds the threshold set by `max_results`
+  # Within each unique grid_id, use the cumsumbinning function from the MESS package
+  # to group sites based on the cumulative sum of results_count across sites that 
+  # share the same characteristic name, resetting the download group/task number if 
+  # the number of records exceeds the threshold set by `max_results`.
   sitecounts_grouped_good_ids <- sitecounts_good_ids %>%
     rename(site_id = MonitoringLocationIdentifier) %>% 
     split(.$grid_id) %>%
     purrr::map_dfr(.f = function(df){
       
       df_grouped <- df %>%
-        arrange(desc(results_count)) %>%
-        mutate(task_num = MESS::cumsumbinning(x = results_count, 
-                                              threshold = max_results, 
-                                              maxgroupsize = max_sites))
+        group_by(CharacteristicName) %>%
+        arrange(desc(results_count), .by_group = TRUE) %>%
+        mutate(task_num_by_results = MESS::cumsumbinning(x = results_count, 
+                                                         threshold = max_results, 
+                                                         maxgroupsize = max_sites), 
+               char_group = cur_group_id(),
+               task_num = NA_integer_) %>%
+        ungroup() 
+      
+      # Each group (which represents a different characteristic name) will have
+      # task numbers that start with "1", so to create unique task numbers, we
+      # modify a column called `task_num` that starts counting the task numbers
+      # where the previous group id left off. For example, both "Specific 
+      # conductance" and "Temperature" may have values for task_num_by_results 
+      # of 1 and 2 but the values of char_group (1 and 2, respectively) would 
+      # mean that they have unique values for task_num equaling 1, 2, 3, and 4.
+      for(i in seq_along(df_grouped$site_id)){
+        
+        if(df_grouped$char_group[i] == 1){
+          df_grouped$task_num[i] <- df_grouped$task_num_by_results[i]
+        } else {
+          max_task_prev_group <- df_grouped %>%
+            filter(char_group == df_grouped$char_group[i] - 1) %>%
+            pull(task_num) %>%
+            max()
+          df_grouped$task_num[i] <- df_grouped$task_num_by_results[i] + max_task_prev_group
```

Comment:
>  Is there any reason that the download groups need to be numeric only?

Not really, the primary reason I didn't concatenate the characteristics with the task numbers was just to print a more interpretable message to the console when downloading the data (i.e., "Fetching data for 11923_1, Specific conductance").

Reply:
Good point. I actually just thought about one other method that may be simple but worth thinking about quickly. Ultimately, the unique combinations of `char_group` and `task_num_by_results` are what we want the task numbers to be. What if we just did a new `group_by()` and then made the `task_num` the `cur_group_id()`? I get the same result doing that:

```r
library(tidyverse)

max_results <- 250000
max_sites <- 500

sitecounts <- tibble(MonitoringLocationIdentifier = c("A","B","C","D","E","F","A","A"), 
                     lon = c(rep(-92.4, 8)), 
                     lat = c(rep(42.5,8)), 
                     datum = c(rep("WGS84",8)), 
                     grid_id = c(rep("11923", 8)), 
                     CharacteristicName = c(rep("Temperature, water", 3), rep("Temperature", 3),"Nitrate","conductivity"), 
                     results_count = c(151, 247000, 7500,15000,235004,134, 249000, 1100))

# This is the code you have in your PR
x <- sitecounts %>%
  rename(site_id = MonitoringLocationIdentifier) %>% 
  split(.$grid_id) %>%
  purrr::map_dfr(.f = function(df){
    
    df_grouped <- df %>%
      group_by(CharacteristicName) %>%
      arrange(desc(results_count)) %>%
      mutate(task_num_by_results = MESS::cumsumbinning(x = results_count, 
                                                       threshold = max_results, 
                                                       maxgroupsize = max_sites), 
             char_group = cur_group_id(),
             task_num = NA_integer_) %>%
      ungroup()
  }) %>% arrange(char_group, task_num_by_results)

x %>% 
  group_by(char_group, task_num_by_results) %>% 
  mutate(task_num = cur_group_id()) %>% 
  ungroup()

# A tibble: 8 x 10
# Groups:   char_group, task_num_by_results [6]
  site_id   lon   lat datum grid_id CharacteristicName results_count task_num_by_results char_group task_num
  <chr>   <dbl> <dbl> <chr> <chr>   <chr>                      <dbl>               <int>      <int>    <int>
1 A       -92.4  42.5 WGS84 11923   conductivity                1100                   1          1        1
2 A       -92.4  42.5 WGS84 11923   Nitrate                   249000                   1          2        2
3 E       -92.4  42.5 WGS84 11923   Temperature               235004                   1          3        3
4 D       -92.4  42.5 WGS84 11923   Temperature                15000                   2          3        4
5 F       -92.4  42.5 WGS84 11923   Temperature                  134                   2          3        4
6 B       -92.4  42.5 WGS84 11923   Temperature, water        247000                   1          4        5
7 C       -92.4  42.5 WGS84 11923   Temperature, water          7500                   2          4        6
8 A       -92.4  42.5 WGS84 11923   Temperature, water           151                   2          4        6
```

Reply:
Ooh, I liked your first suggestion for replacing the for loop, but I think I like this one even better because it's still simple but yet we don't have to mess with the `sitecounts_bad_ids` naming scheme! I've implemented these changes in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/92/commits/e2130c091c470e4d5129cee6f1e15570de078f24.